language/vm/vm-runtime/vm-runtime-types/src/value.rs

```diff
@@ -1043,52 +1046,131 @@ impl Value {
+                        return Err(A::Error::invalid_length(i, &self));
+                    }
+                }
+                Ok(Struct::new(val))
+            }
+        }
+
+        match self {
+            StructDef::Struct(s) => {
+                let fields = s.field_definitions();
+                Ok(Value::struct_(
+                    deserializer.deserialize_tuple(fields.len(), StructVisitor(fields))?,
+                ))
+            }
+            StructDef::Native(ty) => Ok(Value::native_struct(ty.deserialize(deserializer)?)),
         }
     }
-    .map_err(|_| vm_error(Location::new(), StatusCode::INVALID_DATA))
 }
 
-fn deserialize_struct(
-    deserializer: &mut lcs::Deserializer<'_>,
-    struct_def: &StructDef,
-) -> VMResult<Value> {
-    match struct_def {
-        StructDef::Struct(s) => {
-            let mut s_vals = Vec::new();
-            for field_type in s.field_definitions() {
-                s_vals.push(deserialize_value(deserializer, field_type)?);
+impl<'de> de::DeserializeSeed<'de> for &Type {
+    type Value = Value;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        use de::Error;
+
+        match self {
+            Type::Bool => bool::deserialize(deserializer).map(Value::bool),
+            Type::U64 => u64::deserialize(deserializer).map(Value::u64),
+            Type::String => VMString::deserialize(deserializer).map(Value::string),
+            Type::ByteArray => ByteArray::deserialize(deserializer).map(Value::byte_array),
+            Type::Address => AccountAddress::deserialize(deserializer).map(Value::address),
+            Type::Struct(s_fields) => s_fields.deserialize(deserializer),
+            Type::Reference(_) | Type::MutableReference(_) | Type::TypeVariable(_) => {
+                // Case TypeVariable is not possible as all type variable has to be materialized
+                // before serialization.
+                Err(D::Error::custom(
+                    VMStatus::new(StatusCode::INVALID_DATA)
+                        .with_message(format!("Value type {:?} not possible", self)),
+                ))
+            }
+        }
+    }
+}
+
+impl<'de> de::DeserializeSeed<'de> for &NativeStructType {
+    type Value = NativeStructValue;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        use de::Error;
+
+        struct NativeVectorVisitor<'a>(&'a Type);
+        impl<'de, 'a> de::Visitor<'de> for NativeVectorVisitor<'a> {
+            type Value = NativeVector;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("NativeVector")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: de::SeqAccess<'de>,
+            {
+                let mut val = Vec::new();
```

Comment:
What part are you against? Not reserving capacity up front? I’m all for changing this back to reserve some capacity but we’ll have to cap it otherwise we’ll have the same oom issue. Maybe we do: `min(len, 256)` or some such?

Reply:
my take is that the fuzzer should be fixed to be honest, but fine keeping it this way....